core/codegen/src/ident.rs

```diff
@@ -0,0 +1,99 @@
+//! A variant of Ident with more ergonomic handling of raw identifiers.
+
+use std::cmp::Ordering;
+
+use devise::syn;
+use self::syn::ext::IdentExt as _;
+use crate::proc_macro2::{Ident as Ident2, Span};
+use quote::ToTokens;
+
+#[repr(transparent)]
+#[derive(Clone, Debug, Eq, Ord)]
+pub struct Ident(Ident2);
```

Comment:
Would it make more sense to store a `String` instead of an `Ident2`? That's internally what the `Ident2` is, and it would allow us to implement `name` without the `unraw` and as `name(&self) -> &str`. The downside, of course, is then we can't have the `Deref -> Ident2`, so if we use lots of methods from `Ident2`, this is probably a bad idea.

Reply:
On a new enough compiler an `Ident2` contains only a `proc_macro::Ident` which is an opaque handle to the compiler. We probably could store a `String`, but then we also need to store a `Span` and this is where I didn't want to duplicate so much logic. There are some invariants that `Ident` (in `proc_macro` or `proc_macro2` or both, I would need to look again) upholds regarding what the name is.

Reply:
It bothers me that effectively all of these functions require a `String` allocation or two due to the way this is implemented. Can we do something so that we don't allocate on each method call unless we _need_ to, i.e, some downstream use requires a `String`?

Reply:
We could use a [`Storage`](https://docs.rs/state/0.4.1/state/struct.Storage.html) and [`Storage::get_or_set()`](https://docs.rs/state/0.4.1/state/struct.Storage.html#method.get_or_set) to allocate the `String` only if needed.